### DISCUSSION TITLE
FuturesUnordered Task Management

### DISCUSSION THREAD

**Discussion by fooxed** [tokio-rs/tokio#2749] [General]
This is not directly related to Tokio, was not getting responses in the [users forum](https://users.rust-lang.org/t/futuresunordered-task-management/46855/8) and someone requested asking here. As mentioned in that thread, if there is a more appropriate place to ask, please feel free to redirect. The below text is just just reproducing original text from that post:

Trying to understand how [FuturesUnordered](https://docs.rs/futures/0.3.5/futures/stream/struct.FuturesUnordered.html) works. Getting tripped up with how the `ready_to_run_queue` interacts with linked list. Specifically, confused by `pending_next_all`, `spin_next_all`, `link`, and `unlink`. Does anyone have a good reference as to how this works? Specific questions include:

- The docs mention under heavy thread contention, a compare-exchange loop could become costly. What does thread contention mean here? 

- How does using `stub` resolve this? Confused as to how using `pending_next_all` allows `next_all` as needed for iterator. When will there be a case where head_all -> stub -> next?

- More generally, would love an explanation about how `ready_to_run_queue` works with `head_all`, `next_all` pointers and `stub`. Particularly for the impl Stream block

**Comment by Darksonn** [comment]
There was some discussion on this topic [here](https://users.rust-lang.org/t/batch-execution-of-futures-in-the-tokio-runtime-or-max-number-of-active-futures-at-a-time/47659).

**Comment by quininer** [comment]
I think FuturesUnordered based on linked list is too complicated, we can simply use Lock + Queue instead, and sometimes it has good performance.

see https://github.com/quininer/taskset
